### ISSUE TITLE
How to modify the learning rate of SGD based on the validation loss of each epoch using LearningRateScheduler?

### ISSUE THREAD

**Issue by Kevinpsk** [fchollet/keras#2962]
Hi All,

I was wondering how to change the learning rate in SGD based on the validation loss at each epoch end using the LearningRateScheduler callback function? 

I have seen in many papers that people reduce the learning rate by a factor of 2 if the validation loss does not decrease by a certain threshold and stop training when validation loss converges by some standard. Can I do this using the LearningRateScheduler function, or maybe I need to write a new callback function? I know Keras is computing the validation loss at the end of each epoch, but how can I get these loss values while the training is in progress? 

Also in the SGD method provided in Keras, there is an option for learning rate decay, which is implemented as this,

`lr = self.lr * (1. / (1. + self.decay * self.iterations))`

Does this 'iteration' count the number of epochs or mini-batches in one training task? 

Thanks for your help in advance.
Cheers,
Shuokai

**Comment by ChristianThomae** [comment]
> maybe I need to write a new callback function

Yes write a custom callback and use the logs parameter of `def on_epoch_end(self, epoch, logs={})`
to access the current validation loss like: `current = logs.get('val_loss')`.
Update the learning rate similar to the LearningRateScheduler callback: https://github.com/fchollet/keras/blob/master/keras/callbacks.py#L402:L406

`self.iterations` is incremented each weight update: `self.updates = [(self.iterations, self.iterations + 1.)]`, which is done per mini-batch.

**Comment by Kevinpsk** [comment]
Hi Christian,

Thanks a lot for your help. I have also figured this out myself. 

Cheers

**Comment by pengpaiSH** [comment]
@Kevinpsk, could you please share your solution about decreasing learning rates when the validation loss stalls?

**Comment by jingweimo** [comment]
Leave a comment @Kevinpsk : The question as above. Could you please share you solution to updating learning rate based on loss information? Thanks!

**Comment by Kevinpsk** [comment]
@pengpaiSH @jingweimo 
Hi guys, sorry for my late reply. 

I originally changed the learning rate based on validation loss directly on Keras SGD source code, but this is not recommended as you will lose your changes when you update or reinstall Keras when something went wrong. So you should follow Christian's suggestion to write a custom callback function. You can find instructions on how to do this in Keras documentation. 

Good luck.

**Comment by blaze-gitter** [comment]
For anyone still ending up here, you might be looking for 

tf.keras.callbacks.ReduceLROnPlateau
